File: data/books/cs5e/ch07/fig0738/Translator.java
package fig0738;

import java.util.ArrayList;

/**
 * The parser implemented as a finite state machine.
 *
 * <p>
 * File: <code>Translator.java</code>
 *
 * @author J. Stanley Warford
 */
public class Translator {

   private final InBuffer b;
   private Tokenizer t;
   private ACode aCode;

   public Translator(InBuffer inBuffer) {
      b = inBuffer;
   }

   // Sets aCode and returns boolean true if end statement is processed.
   private boolean parseLine() {
      boolean terminate = false;
      AArg localFirstArg = new IntArg(0);
      AArg localSecondArg;
      Mnemon localMnemon = Mnemon.M_END; // Useless initialization
      AToken aToken;
      aCode = new EmptyInstr();
      ParseState state = ParseState.PS_START;
      do {
         aToken = t.getToken();
         switch (state) {
            case PS_START:
               if (aToken instanceof TIdentifier) {
                  TIdentifier localTIdentifier = (TIdentifier) aToken;
                  String tempStr = localTIdentifier.getStringValue();
                  if (Maps.unaryMnemonTable.containsKey(
                        tempStr.toLowerCase())) {
                     localMnemon = Maps.unaryMnemonTable.get(
                           tempStr.toLowerCase());
                     aCode = new UnaryInstr(localMnemon);
                     terminate = localMnemon == Mnemon.M_END;
                     state = ParseState.PS_UNARY;
                  } else if (Maps.nonUnaryMnemonTable.containsKey(
                        tempStr.toLowerCase())) {
                     localMnemon = Maps.nonUnaryMnemonTable.get(
                           tempStr.toLowerCase());
                     state = ParseState.PS_FUNCTION;
                  } else {
                     aCode = new Error(
                           "Line must begin with function identifier.");
                  }
               } else if (aToken instanceof TEmpty) {
                  aCode = new EmptyInstr();
                  state = ParseState.PS_FINISH;
               } else {
                  aCode = new Error(
                        "Line must begin with function identifier.");
               }
               break;
            case PS_FUNCTION:
               if (aToken instanceof TLeftParen) {
                  state = ParseState.PS_OPEN;
               } else {
                  aCode = new Error(
                        "Left parenthesis expected after function.");
               }
               break;
            case PS_OPEN:
               if (aToken instanceof TIdentifier) {
                  TIdentifier localTIdentifier = (TIdentifier) aToken;
                  localFirstArg = new IdentArg(
                        localTIdentifier.getStringValue());
                  state = ParseState.PS_1ST_OPRND;
               } else {
                  aCode = new Error("First argument not an identifier.");
               }
               break;
            case PS_1ST_OPRND:
               if (localMnemon == Mnemon.M_NEG
                     || localMnemon == Mnemon.M_ABS) {
                  if (aToken instanceof TRightParen) {
                     aCode = new OneArgInstr(localMnemon, localFirstArg);
                     state = ParseState.PS_NON_UNARY1;
                  } else {
                     aCode = new Error(
                           "Right parenthesis expected after argument.");
                  }
               } else if (aToken instanceof TComma) {
                  state = ParseState.PS_COMMA;
               } else {
                  aCode = new Error(
                        "Comma expected after first argument.");
               }
               break;
            case PS_COMMA:
               if (aToken instanceof TIdentifier) {
                  TIdentifier localTIdentifier = (TIdentifier) aToken;
                  localSecondArg = new IdentArg(
                        localTIdentifier.getStringValue());
                  aCode = new TwoArgInstr(
                        localMnemon, localFirstArg, localSecondArg);
                  state = ParseState.PS_2ND_OPRND;
               } else if (aToken instanceof TInteger) {
                  TInteger localTInteger = (TInteger) aToken;
                  localSecondArg = new IntArg(
                        localTInteger.getIntValue());
                  aCode = new TwoArgInstr(
                        localMnemon, localFirstArg, localSecondArg);
                  state = ParseState.PS_2ND_OPRND;
               } else {
                  aCode = new Error(
                        "Second argument not an identifier or integer.");
               }
               break;
            case PS_2ND_OPRND:
               if (aToken instanceof TRightParen) {
                  state = ParseState.PS_NON_UNARY2;
               } else {
                  aCode = new Error(
                        "Right parenthesis expected after argument.");
               }
               break;
            case PS_UNARY:
            case PS_NON_UNARY1:
            case PS_NON_UNARY2:
               if (aToken instanceof TEmpty) {
                  state = ParseState.PS_FINISH;
               } else {
                  aCode = new Error("Illegal trailing character.");
               }
               break;
         }
      } while (state != ParseState.PS_FINISH && !(aCode instanceof Error));
      return terminate;
   }

   public void translate() {
      ArrayList<ACode> codeTable = new ArrayList<>();
      int numErrors = 0;
      t = new Tokenizer(b);
      boolean terminateWithEnd = false;
      b.getLine();
      while (b.inputRemains() && !terminateWithEnd) {
         terminateWithEnd = parseLine(); // Sets aCode and returns boolean.
         codeTable.add(aCode);
         if (aCode instanceof Error) {
            numErrors++;
         }
         b.getLine();
      }
      if (!terminateWithEnd) {
         aCode = new Error("Missing \"end\" sentinel.");
         codeTable.add(aCode);
         numErrors++;
      }
      if (numErrors == 0) {
         System.out.print("Object code:\n");
         for (ACode code : codeTable) {
            System.out.printf("%s", code.generateCode());
         }
      }
      if (numErrors == 1) {
         System.out.print("One error was detected.\n");
      } else if (numErrors > 1) {
         System.out.printf("%d errors were detected.\n", numErrors);
      }
      System.out.print("\nProgram listing:\n");
      for (ACode code : codeTable) {
         System.out.printf("%s", code.generateListing());
      }
   }
}

File: data/books/cs6e/ch07/fig0738/translator.py
import enum
import io
from typing import List, cast

from .arguments import IdentifierArgument, IntegerArgument
from .code import ACode, Error, UnaryInstr
from .code import EmptyInstr, OneArgInstr, TwoArgInstr
from .mnemonics import Mnemonics
from .tokenizer import Tokenizer
from .tokens import Tokens


class Translator:
    class States(enum.Enum):
        START = 1
        UNARY = 2
        FUNCTION = 3
        OPEN = 4
        OPERAND_1ST = 5
        OPERAND_2ND = 6
        NONUNARY1 = 7
        NONUNARY2 = 8
        COMMA = 9
        FINISH = 10

    def __init__(self, buffer: io.StringIO):
        self.buffer = buffer
        self.tokenizer = Tokenizer(self.buffer)

    def parse_line(self) -> ACode | None:
        state = Translator.States.START
        code, arg1, arg2 = None, None, None
        mnemonic: Mnemonics = Mnemonics.END
        while state != Translator.States.FINISH and type(code) is not Error:
            token_type, token_val = self.tokenizer.next_token()
            match state:
                case Translator.States.START:
                    if token_type == Tokens.IDENTIFIER:
                        as_upper = token_val.upper()
                        try:
                            as_mnemonic: Mnemonics = Mnemonics[as_upper]
                            if as_mnemonic.is_unary():
                                code = UnaryInstr(as_mnemonic)
                                state = Translator.States.UNARY
                            else:
                                mnemonic = as_mnemonic
                                state = Translator.States.FUNCTION
                        except KeyError:
                            code = Error("Line must begin with function identifier")
                    elif token_type == Tokens.EMPTY:
                        code = EmptyInstr()
                        state = Translator.States.FINISH
                    else:
                        code = Error("Line must begin with function identifier")

                case Translator.States.FUNCTION:
                    if token_type == Tokens.LEFT_PAREN:
                        state = Translator.States.OPEN
                    else:
                        code = Error("Left parenthesis expected after function.")

                case Translator.States.OPEN:
                    if token_type == Tokens.IDENTIFIER:
                        arg1 = IdentifierArgument(token_val)
                        state = Translator.States.OPERAND_1ST
                    else:
                        code = Error("First argument is not an identifier.")

                case Translator.States.OPERAND_1ST:
                    if mnemonic in {Mnemonics.NEG, Mnemonics.ABS}:
                        if token_type == Tokens.RIGHT_PAREN:
                            code = OneArgInstr(mnemonic, arg1)
                            state = Translator.States.NONUNARY1
                        else:
                            code = Error("Right parenthesis expected after argument.")
                    elif token_type == Tokens.COMMA:
                        state = Translator.States.COMMA
                    else:
                        code = Error("Comma expected after first argument.")

                case Translator.States.COMMA:
                    if token_type == Tokens.IDENTIFIER:
                        arg2 = IdentifierArgument(token_val)
                        state = Translator.States.OPERAND_2ND
                    elif token_type == Tokens.INTEGER:
                        arg2 = IntegerArgument(token_val)
                        state = Translator.States.OPERAND_2ND
                    else:
                        code = Error("Second argument not an identifier or integer.")

                case Translator.States.OPERAND_2ND:
                    if token_type == Tokens.RIGHT_PAREN:
                        code = TwoArgInstr(mnemonic, arg1, arg2)
                        state = Translator.States.NONUNARY2
                    else:
                        code = Error("Right parenthesis expected after argument.")

                # Unary or NonUnary1 or NonUnary2
                case _:
                    if token_type == Tokens.EMPTY:
                        state = Translator.States.FINISH
                    else:
                        code = Error("Illegal trailing character.")

        return code

    def translate(self):
        program: List[ACode] = []
        error_count = 0

        while (code := self.parse_line()) is not None:
            program.append(code)
            if type(code) is Error:
                error_count += 1
            # TODO: ugly hack that should be a state variable instead. Will revisit when this becomes iterable.
            elif (
                type(code) is UnaryInstr
                and cast(code, UnaryInstr).mnemonic == Mnemonics.END
            ):
                break
        else:  # No break
            program.append(Error('Missing "end" sentinel'))
            error_count += 1

        if error_count == 0:
            print("Object code:")
            print(
                *(code := line.generate_code() for line in program if code != ""),
                sep="",
            )
        elif error_count == 1:
            print("One error was detected.")
        else:
            print(f"{error_count} errors were detected.")

        print("Program listing:")
        print(*(line.generate_listing() for line in program), sep="")
